src/yb/yql/cql/cqlserver/system_query_cache.cc

```diff
@@ -0,0 +1,186 @@
+
+  "SELECT * FROM system.local WHERE key='local'",
+  "SELECT schema_version FROM system.local WHERE key='local'",
+  // The client evidently doesn't always have consistent formatting
+  "select * from system.local where key = 'local'",
+  "select schema_version from system.local where key='local'",
+  "SELECT * FROM system_schema.keyspaces",
+  "SELECT * FROM system_schema.tables",
+  "SELECT * FROM system_schema.views",
+  "SELECT * FROM system_schema.columns",
+  "SELECT * FROM system_schema.types",
+  "SELECT * FROM system_schema.functions",
+  "SELECT * FROM system_schema.aggregates",
+  "SELECT * FROM system_schema.triggers",
+  "SELECT * FROM system_schema.indexes",
+};
+
+SystemQueryCache::SystemQueryCache(cqlserver::CQLServiceImpl* service_impl)
+  : service_impl_(service_impl) {
+  pool_ = std::make_unique<yb::rpc::IoThreadPool>("system_query_cache_updater", 1);
+  scheduler_ = std::make_unique<yb::rpc::Scheduler>(&pool_->io_service());
+
+  LOG(INFO) << "Created system cache updater.";
+
+  if (FLAGS_cql_update_system_cache_msecs > 0) {
+    if (MonoDelta::FromSeconds(FLAGS_cql_system_cache_stale_secs) <
+        MonoDelta::FromMilliseconds(FLAGS_cql_update_system_cache_msecs)) {
+      LOG(WARNING) << "Stale expiration shorter than update rate.";
+    }
+
+    ScheduleRefreshCache(true /* now */);
+  } else {
+    LOG(WARNING) << "System cache created with nonpositive timeout. Disabling scheduling.";
+  }
+}
+
+SystemQueryCache::~SystemQueryCache() {
+  if (pool_) {
+    scheduler_->Shutdown();
+    pool_->Shutdown();
+    pool_->Join();
+  }
+}
+
+boost::optional<RowsResult::SharedPtr> SystemQueryCache::Lookup(const std::string& query) {
+  if (FLAGS_cql_system_cache_stale_secs > 0 &&
+      GetStaleness() > MonoDelta::FromSeconds(FLAGS_cql_system_cache_stale_secs)) {
+    return boost::none;
+  }
+  const std::lock_guard<std::mutex> l(cache_mutex_);
+
+  const auto it = cache_.find(query);
+  if (it == cache_.end()) {
+    return boost::none;
+  } else {
+    return it->second;
+  }
+}
+
+MonoDelta SystemQueryCache::GetStaleness() {
+  const std::lock_guard<std::mutex> l(cache_mutex_);
+  return MonoTime::Now() - last_updated_;
+}
+
+void SystemQueryCache::RefreshCache() {
+  VLOG(1) << "Refreshing system query cache";
+  std::unordered_map<std::string, RowsResult::SharedPtr> new_cache;
+  for (auto query : SYSTEM_QUERIES) {
+    Status status;
+    ExecutedResult::SharedPtr result;
+    ExecuteSync(query, &status, &result);
+
+    if (status.ok()) {
+      new_cache[query] = std::dynamic_pointer_cast<RowsResult>(result);
+    } else {
+      LOG(WARNING) << "Could not execute statement: " << query;
+      // We don't want to update the cache with no data; instead we'll let the
+      // stale cache persist.
+      ScheduleRefreshCache(false /* now */);
+      return;
+    }
+  }
+
+  {
+    const std::lock_guard<std::mutex> l(cache_mutex_);
+    // I'd like to do `cache_ = new_cache`, but for some reason that's not allowed.
```

Comment:
It's better if you make cache_ to be a shared_ptr (unique_ptr might be better) to the unordered map and create the object it points to in the constructor. And new_cache can be defined like:
```
auto new_cache = make_shared<std::unordered_map<std::string, RowsResult::SharedPtr>>();
```

Then you can reset the cache_ pointer atomically while holding the lock